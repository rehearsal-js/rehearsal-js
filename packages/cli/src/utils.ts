import { join, resolve } from 'path';
import { readFileSync, writeJSONSync, readJSONSync } from 'fs-extra';
import { compare } from 'compare-versions';
import { parse } from 'json5';
import { valid } from 'semver';
import { type SimpleGit, type SimpleGitOptions, simpleGit } from 'simple-git';
import which from 'which';
import { InvalidArgumentError } from 'commander';

import findup = require('findup-sync');
import execa = require('execa');

import { ScriptMap } from './types';
import type { GitDescribe } from './interfaces';

export const VERSION_PATTERN = /_(\d+\.\d+\.\d+)/;

export const git: SimpleGit = simpleGit({
  baseDir: process.cwd(),
  binary: 'git',
  maxConcurrentProcesses: 6,
} as Partial<SimpleGitOptions>);

// we might need to check if commitlint is being used and lint against it for msgTypes
export async function gitCommit(msg: string, msgType = 'chore(deps-dev)'): Promise<void> {
  await git.commit(`${msgType.trim()}: [REHEARSAL-BOT] ${msg.trim()}`, '--no-verify');
}

// we want a seperate branch & PR for each diagnostic
// eg. await gitCheckoutNewBranch('4.8.0-beta', 'typescript', 'diagnostic-4082')
export async function gitCheckoutNewLocalBranch(
  depSemVersion: string,
  depName = 'typescript',
  addPathID?: string
): Promise<string> {
  const branchName = `rehearsal-bot/${depName}-${depSemVersion}${addPathID ? `/${addPathID}` : ''}`;
  await git.checkout(['-b', branchName]);

  return branchName;
}

export async function gitIsRepoDirty(cwd?: string): Promise<boolean> {
  // if cwd is provided, create a new git instance with the correct cwd
  const gitClient = cwd
    ? simpleGit({
        baseDir: cwd,
        binary: 'git',
        maxConcurrentProcesses: 6,
      } as Partial<SimpleGitOptions>)
    : git;
  if (await gitClient.checkIsRepo()) {
    const status = await gitClient.status();
    return status.isClean() === false;
  }
  return false; // false if it's not a git repo
}

/**
 * Function to introduce a wait
 *
 * @param ms - how many milliseconds to wait
 */
export function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

/**
 * Convert milliseconds to seconds
 *
 * @param ms - number in milliseconds
 * @returns number in seconds
 */
export function msToSeconds(ms: number): number {
  return Math.round(ms / 1000);
}

export function readJSON<T>(file: string): T | undefined {
  const text = readText(file);
  if (text !== undefined) {
    return parse(text);
  }
}

export function readText(file: string): string | undefined {
  return readFile(file, 'utf8');
}

export function readFile(file: string, encoding: 'utf8'): string | undefined;
export function readFile(file: string, encoding?: undefined): Buffer | undefined;
export function readFile(file: string, encoding?: 'utf8'): string | Buffer | undefined {
  try {
    return readFileSync(file, encoding);
  } catch (e) {
    if ((e as NodeJS.ErrnoException).code === 'ENOENT') {
      return;
    }
    throw e;
  }
}

/**
 * Provide current timestamp
 * @param inSeconds - if true, return seconds, otherwise milliseconds
 * @returns timestamp in seconds
 */
export function timestamp(inSeconds = false): number {
  return inSeconds ? Date.now() / 1000 : Date.now();
}

/**
 * Parse the string output from git
 *
 * @param desc - String output of "git describe --tags --long"
 * @returns GitDescribe object
 */
export function parseLongDescribe(desc: string): GitDescribe {
  const result = /^(.+)-(\d+)-g([a-f0-9]+)(?:-(dirty))?$/.exec(desc);

  if (!result) {
    throw Error(`Unable to parse ${desc} as a long description`);
  }

  const [, tag, count, sha, dirty] = result;

  return {
    tag,
    count: parseInt(count, 10),
    sha,
    dirty: !!dirty,
  };
}

/**
 * Convert instances like "foo-web_10.2.3" and "1.2.3" to "1.2.3"
 *
 * @param versionString - version string to convert into just numbers and periods
 * @returns version string with only numbers and periods
 */
export function normalizeVersionString(versionString: string): string {
  if (VERSION_PATTERN.test(versionString)) {
    const result = VERSION_PATTERN.exec(versionString);
    return result ? result[1] : versionString;
  }
  return versionString;
}

export function determineProjectName(directory = process.cwd()): string | null {
  const packageJSONPath = findup('package.json', {
    cwd: directory,
  });

  if (!packageJSONPath) {
    return null;
  }
  const packageJSON = readJSON<{ name: string }>(packageJSONPath);
  return packageJSON?.name ?? null;
}

export function isYarnManager(): boolean {
  const yarnPath = findup('yarn.lock', {
    cwd: process.cwd(),
  });

  return !!yarnPath;
}

export function isYarnBerryManager(): boolean {
  const lockFilePath = findup('yarn.lock', {
    cwd: process.cwd(),
  });
  const berryConfigPath = findup('.yarnrc.yml', {
    cwd: process.cwd(),
  });

  return !!lockFilePath && !!berryConfigPath;
}

export function isPnpmManager(): boolean {
  const pnpmPath = findup('pnpm-lock.yaml', {
    cwd: process.cwd(),
  });

  return !!pnpmPath;
}

export function getLockfilePath(): string | null {
  const yarnPath = findup('yarn.lock', {
    cwd: process.cwd(),
  });

  if (yarnPath) {
    return yarnPath;
  }

  const pnpmPath = findup('pnpm-lock.yaml', {
    cwd: process.cwd(),
  });

  if (pnpmPath) {
    return pnpmPath;
  }

  const npmPath = findup('package-lock.json', {
    cwd: process.cwd(),
  });

  if (npmPath) {
    return npmPath;
  }

  return null;
}

export function getModuleManager(): 'yarn' | 'npm' | 'pnpm' {
  const isYarn = isYarnManager();
  const isPnpm = isPnpmManager();

  if (isYarn) {
    return 'yarn';
  }

  if (isPnpm) {
    return 'pnpm';
  }

  return 'npm';
}

export function getModuleManagerInstaller(
  manager: 'yarn' | 'npm' | 'pnpm',
  depList: string[],
  isDev: boolean
): { bin: string; args: string[] } {
  const bin = which.sync(manager);

  switch (manager) {
    case 'yarn':
      // since yarn@3 doesn't have --ignore-scripts anymore
      return {
        bin,
        args: isDev
          ? ['add', '-D', ...depList, ...(isYarnBerryManager() ? [] : ['--ignore-scripts'])]
          : ['add', ...depList, ...(isYarnBerryManager() ? [] : ['--ignore-scripts'])],
      };
    case 'pnpm':
      return {
        bin,
        args: isDev ? ['add', '-D', ...depList] : ['add', ...depList],
      };
    case 'npm':
    default:
      return {
        bin,
        args: isDev
          ? ['install', ...depList, '--save-dev', '--ignore-scripts']
          : ['install', ...depList, '--ignore-scripts'],
      };
  }
}

// devDep: dep@version eg. typescript@4.4.4 or typescript or etc
export async function addDep(
  depList: string[],
  isDev: boolean,
  options: execa.Options = {}
): Promise<void> {
  const moduleManager = getModuleManager();
  const binAndArgs = getModuleManagerInstaller(moduleManager, depList, isDev);

  await execa(binAndArgs.bin, [...binAndArgs.args], options);
}

// when executing a command with module manager prefix
// eg. yarn tsc or pnpm tsc or npm tsc
export async function runModuleCommand(args: string[], option: execa.Options = {}): Promise<void> {
  const moduleManager = getModuleManager();
  const binAndArgs = {
    bin: moduleManager,
    args,
  };
  await execa(binAndArgs.bin, [...binAndArgs.args], option);
}

// rather than explicitly setting from node_modules dir we need to handle workspaces use case
// and volta use case
export async function getPathToBinary(
  binaryName: string,
  options: execa.Options = {}
): Promise<string> {
  // pnpm | yarn | npm
  const moduleManager = getModuleManager();
  // /Users/foo/.volta/bin/yarn
  // /usr/local/bin/pnpm
  const moduleManagerBin = which.sync(moduleManager);

  let stdoutMsg;

  try {
    const { stdout } = await execa(moduleManagerBin, ['bin', binaryName], options);
    stdoutMsg = stdout.trim().split(`/${binaryName}`)[0];
  } catch (e) {
    throw new Error(`Unable to find binary path to ${binaryName}`);
  }

  // return the path to the binary
  try {
    return resolve(join(stdoutMsg, binaryName));
  } catch (error) {
    throw new Error(`Unable to find ${binaryName} with ${moduleManagerBin}`);
  }
}

// handles instances where latest is newer than beta and rc etc.
// latestBeta = whichever is latest between latest, beta and rc tags
export async function getLatestTSVersion(tag: string): Promise<string> {
  const module = 'typescript';

  if (tag === 'latestBeta') {
    const compareTags = ['latest', 'beta', 'rc'];
    const versions = await Promise.all(compareTags.map((t) => getLatestModuleVersion(module, t)));
    // compare versions and return the latest
    return versions.reverse().reduce((acc, curr) => {
      if (compare(curr, acc, '>')) {
        return curr;
      }
      return acc;
    }, '0.0.0');
  } else {
    const { stdout } = await execa('npm', ['show', `typescript@${tag}`, 'version']);
    return stdout;
  }
}

export async function getLatestModuleVersion(module: string, tag: string): Promise<string> {
  const { stdout } = await execa('npm', ['show', `${module}@${tag}`, 'version']);

  return stdout;
}

export function isValidSemver(input: string): boolean {
  return !!valid(input);
}

/**
 * Parses the comma separated string `a,b,c` to an array os strings [a, b, c]
 */
export function parseCommaSeparatedList(value: string): string[] {
  return value.split(',');
}

/**
 * Generate tsconfig
 */
export function writeTSConfig(basePath: string, fileList: string[]): void {
  const include = [...fileList.map((f) => f.replace('.js', '.ts'))];
  const config = {
    $schema: 'https://json.schemastore.org/tsconfig',
    compilerOptions: {
      strict: true,
      esModuleInterop: true,
      noUncheckedIndexedAccess: true,
      module: 'es2020',
      moduleResolution: 'node',
      newLine: 'lf',
      target: 'ES2021',
      forceConsistentCasingInFileNames: true,
      noFallthroughCasesInSwitch: true,
      noEmit: true,
    },
    include,
  };

  writeJSONSync(resolve(basePath, 'tsconfig.json'), config, { spaces: 2 });
}

/**
 * Parse the TS Version passed in by the user
 */
export function parseTsVersion(value: string): string {
  if (isValidSemver(value)) {
    return value;
  } else {
    throw new InvalidArgumentError(
      'The tsVersion specified is an invalid string. Please specify a valid version as n.n.n'
    );
  }
}

/**
 * Check if typescript is already installed
 */
export function isTypescriptInDevdep(basePath: string): boolean {
  const packageJSONPath = resolve(basePath, 'package.json');
  const packageJSON = readJSONSync(packageJSONPath);
  return (
    (packageJSON.devDependencies && packageJSON.devDependencies.typescript) ||
    (packageJSON.dependencies && packageJSON.dependencies.typescript)
  );
}

/**
 * Add/Update scripts in package.json
 */
export function addPackageJsonScripts(basePath: string, scriptMap: ScriptMap): void {
  const packageJSONPath = resolve(basePath, 'package.json');
  const packageJSON = readJSONSync(packageJSONPath);
  packageJSON.scripts = { ...packageJSON.scripts, ...scriptMap };
  writeJSONSync(packageJSONPath, packageJSON, { spaces: 2 });
}

/**
 * Run git reset to get all file changes to the previous state
 */
export async function resetFiles(): Promise<void> {
  if (await git.checkIsRepo()) {
    await git.reset(['--hard']);
  }
}
