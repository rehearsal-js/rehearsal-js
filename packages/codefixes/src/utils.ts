import { getLineAndCharacterOfPosition } from 'typescript';
import { CodeFixAction, FixedFile } from './types';
import type { SourceFile } from 'typescript';

export function getCodemodData(
  modifiedSourceFile: SourceFile,
  updatedText: string,
  insertionPos: number,
  code: string,
  action: CodeFixAction
): FixedFile[] {
  const { line, character } = getLineAndCharacterOfPosition(modifiedSourceFile, insertionPos);
  const startLine = line + 1; //bump line 0 to line 1, so on and so forth
  const startColumn = character + 1; //bump character 0 to character 1, so on and so forth

  return [
    {
      fileName: modifiedSourceFile.fileName,
      updatedText,
      code: code,
      location: {
        startLine,
        startColumn,
        endLine: startLine, //TODO: calculate endLine for multiple line insertion
        endColumn: getEndColumn(startColumn, code, action),
      },
      codeFixAction: action,
    },
  ];
}

function getEndColumn(startColumn: number, code: string, action: CodeFixAction): number {
  switch (action) {
    case 'add':
    case 'replace':
      return startColumn + code.length;
      break;
    case 'delete':
      return startColumn;
      break;
  }
}
