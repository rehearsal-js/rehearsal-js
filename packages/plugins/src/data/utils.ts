import type { FixedFile } from '@rehearsal/codefixes';
import type { FileRole, Location, ProcessedFile } from '@rehearsal/reporter';
import type { DiagnosticWithLocation, SourceFile } from 'typescript';
import { getLineAndCharacterOfPosition } from 'typescript';

export function getFilesData(
  fixedFiles: FixedFile[],
  diagnostic: DiagnosticWithLocation,
  hint = ''
): { [fileName: string]: ProcessedFile } {
  const entryFileName = diagnostic.file.fileName;

  let filesData: { [fileName: string]: ProcessedFile } = {
    [entryFileName]: getInitialEntryFileData(diagnostic),
  };

  if (fixedFiles.length > 0) {
    for (const fixedFile of fixedFiles) {
      filesData = {
        ...filesData,
        [fixedFile.fileName]: getFixedFileData(fixedFile, entryFileName),
      };
    }
  } else {
    filesData = {
      [entryFileName]: {
        ...filesData[entryFileName],
        hintAdded: true,
        hint,
      },
    };
  }

  return filesData;
}

function getRoles(fileName: string, entryFileName: string, fixed: boolean): FileRole[] {
  const isEntryFile = fileName === entryFileName;

  if (isEntryFile) {
    return fixed
      ? ['analysisTarget' as const, 'modified' as const]
      : ['analysisTarget', 'unmodified'];
  }
  return fixed ? ['tracedFile' as const, 'modified' as const] : ['tracedFile', 'unmodified'];
}

export function getLocation(sourceFile: SourceFile, start: number): Location {
  const { line, character } = getLineAndCharacterOfPosition(sourceFile, start);

  return {
    line: line + 1,
    character: character + 1,
  };
}

export function getFixedFileData(fixedFile: FixedFile, entryFileName: string): ProcessedFile {
  const roles = getRoles(fixedFile.fileName, entryFileName, true);
  return {
    fileName: fixedFile.fileName,
    location: fixedFile.location,
    fixed: true,
    code: fixedFile.code,
    codeFixAction: fixedFile.codeFixAction,
    hint: undefined,
    hintAdded: false,
    roles,
  };
}

export function getCommentedFileData(
  fileName: string,
  location: Location,
  hint: string,
  entryFileName: string
): ProcessedFile {
  const roles = getRoles(fileName, entryFileName, false);
  return {
    fileName,
    location,
    fixed: false,
    code: undefined,
    codeFixAction: undefined,
    hint: hint,
    hintAdded: true,
    roles,
  };
}

export function getInitialEntryFileData(diagnostic: DiagnosticWithLocation): ProcessedFile {
  return {
    fileName: diagnostic.file.fileName,
    location: getLocation(diagnostic.file, diagnostic.start),
    fixed: false,
    code: undefined,
    codeFixAction: undefined,
    hint: undefined,
    hintAdded: false,
    roles: ['analysisTarget', 'unmodified'],
  };
}
