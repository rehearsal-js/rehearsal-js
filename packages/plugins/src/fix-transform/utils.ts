import ts from 'typescript';
import { type FixedFile } from '.';

export function getCodemodResult(
  modifiedSourceFile: ts.SourceFile,
  updatedText: string,
  insertionPos: number
): FixedFile[] {
  const { line, character } = ts.getLineAndCharacterOfPosition(modifiedSourceFile, insertionPos);
  return [
    {
      fileName: modifiedSourceFile.fileName,
      updatedText,
      location: {
        line,
        character,
      },
    },
  ];
}
