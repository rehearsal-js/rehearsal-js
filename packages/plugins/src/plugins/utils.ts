import ts from 'typescript';

/**
 * The following logic has been ported from https://github.com/airbnb/ts-migrate/blob/master/packages/ts-migrate-plugins/src/plugins/ts-ignore.ts
 **/

export function onMultilineConditionalTokenLine(sourceFile: ts.SourceFile, pos: number): boolean {
  const conditionalExpression = getConditionalExpressionAtPos(sourceFile, pos);
  // Not in a conditional expression.
  if (!conditionalExpression) {
    return false;
  }

  const { line: questionTokenLine } = ts.getLineAndCharacterOfPosition(
    sourceFile,
    conditionalExpression.questionToken.end
  );
  const { line: colonTokenLine } = ts.getLineAndCharacterOfPosition(
    sourceFile,
    conditionalExpression.colonToken.end
  );
  // Single line conditional expression.
  if (questionTokenLine === colonTokenLine) {
    return false;
  }

  const { line } = ts.getLineAndCharacterOfPosition(sourceFile, pos);
  return visitConditionalExpressionWhen(conditionalExpression, pos, {
    // On question token line of multiline conditional expression.
    whenTrue: () => line === questionTokenLine,
    // On colon token line of multiline conditional expression.
    whenFalse: () => line === colonTokenLine,
    otherwise: () => false,
  });
}

export function visitConditionalExpressionWhen<T>(
  node: ts.ConditionalExpression | undefined,
  pos: number,
  visitor: {
    whenTrue(node: ts.ConditionalExpression): T;
    whenFalse(node: ts.ConditionalExpression): T;
    otherwise(): T;
  }
): T {
  if (!node) {
    return visitor.otherwise();
  }

  const inWhenTrue = node.whenTrue.pos <= pos && pos < node.whenTrue.end;
  if (inWhenTrue) {
    return visitor.whenTrue(node);
  }

  const inWhenFalse = node.whenFalse.pos <= pos && pos < node.whenFalse.end;
  if (inWhenFalse) {
    return visitor.whenFalse(node);
  }

  return visitor.otherwise();
}

export function getConditionalExpressionAtPos(
  sourceFile: ts.SourceFile,
  pos: number
): ts.ConditionalExpression | undefined {
  const visitor = (node: ts.Node): ts.ConditionalExpression | undefined => {
    if (node.pos <= pos && pos < node.end && ts.isConditionalExpression(node)) {
      return node;
    }
    return ts.forEachChild(node, visitor);
  };
  return ts.forEachChild(sourceFile, visitor);
}

export function getConditionalCommentPos(sourceFile: ts.SourceFile, pos: number): number {
  return visitConditionalExpressionWhen(getConditionalExpressionAtPos(sourceFile, pos), pos, {
    whenTrue: (node) => node.questionToken.end,
    whenFalse: (node) => node.colonToken.end,
    otherwise: () => pos,
  });
}

export function inTemplateExpressionText(sourceFile: ts.SourceFile, pos: number): boolean {
  const visitor = (node: ts.Node): boolean | undefined => {
    if (node.pos <= pos && pos < node.end && ts.isTemplateExpression(node)) {
      const inHead = node.head.pos <= pos && pos < node.head.end;
      const inMiddleOrTail = node.templateSpans.some(
        (span) => span.literal.pos <= pos && pos < span.literal.end
      );
      if (inHead || inMiddleOrTail) {
        return true;
      }
    }

    return ts.forEachChild(node, visitor);
  };

  return !!ts.forEachChild(sourceFile, visitor);
}

export function findDiagnosticNode(
  diagnostic: ts.DiagnosticWithLocation,
  sourceFile: ts.SourceFile
): ts.Node | undefined {
  const visitor = (node: ts.Node): ts.Node | undefined =>
    isDiagnosticNode(node, diagnostic, sourceFile) ? node : ts.forEachChild(node, visitor);

  return visitor(sourceFile);
}

export function isDiagnosticNode(
  node: ts.Node,
  diagnostic: ts.DiagnosticWithLocation,
  sourceFile: ts.SourceFile
): boolean {
  return (
    node.getStart(sourceFile) === diagnostic.start &&
    node.getEnd() === diagnostic.start + diagnostic.length
  );
}

export function inJsxText(sourceFile: ts.SourceFile, pos: number): boolean {
  const visitor = (node: ts.Node): boolean | undefined => {
    if (node.pos <= pos && pos < node.end && (ts.isJsxElement(node) || ts.isJsxFragment(node))) {
      const isJsxTextChild = node.children.some(
        (child) => ts.isJsxText(child) && child.pos <= pos && pos < child.end
      );
      const isClosingElement = !ts.isJsxFragment(node) && node.closingElement.pos === pos;
      if (isJsxTextChild || isClosingElement) {
        return true;
      }
    }

    return ts.forEachChild(node, visitor);
  };

  return !!ts.forEachChild(sourceFile, visitor);
}

export function getBoundaryOfCommentBlock(
  start: number,
  length: number,
  text: string
): { start: number; end: number } {
  const newStart = start - 1 >= 0 && text[start - 1] === '{' ? start - 1 : start;

  let end = start + length - 1;

  end = end + 1 < text.length && text[end + 1] === '}' ? end + 1 : end;
  end = ts.isLineBreak(text.charCodeAt(end + 1)) ? end + 1 : end;

  return {
    start: newStart,
    end,
  };
}
