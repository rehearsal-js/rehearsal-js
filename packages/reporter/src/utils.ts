import { readFileSync } from 'fs';
import { parse } from 'json5';

export function readText(file: string): string | undefined {
  return readFile(file, 'utf8');
}

export function readFile(file: string, encoding: 'utf8'): string | undefined;
export function readFile(file: string, encoding?: undefined): Buffer | undefined;
export function readFile(file: string, encoding?: 'utf8'): string | Buffer | undefined {
  try {
    return readFileSync(file, encoding);
  } catch (e) {
    if ((e as NodeJS.ErrnoException).code === 'ENOENT') {
      return;
    }
    throw e;
  }
}

export function readJSONString<TJson = unknown>(filepath: string, separator = '\n'): TJson[] {
  const text = readText(filepath) as string;
  const strJSONArray = text.split(separator).filter((a) => a);

  return strJSONArray.map((a) => parse(a));
}

export async function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}
