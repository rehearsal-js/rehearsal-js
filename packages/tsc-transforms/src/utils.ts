// replaces the positional contents in a string
// eg. "The declaration '{0}' is never read. Remove the declaration or use it."
// '{0}' is replaced with the first argument from the replacements array
export function strPositionalReplacement(message: string, replacements: string[]): string {
  replacements.map((r, i) => {
    // strip single quotes and excess whitespace from the replacements
    const posArg = r.replace(/'/gi, '').trim();
    // insert the positional arguments from the help message
    message = message.replace(`{${i}}`, posArg);
  });

  return message;
}
