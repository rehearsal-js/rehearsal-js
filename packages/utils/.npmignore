dist/src/__tests__
dist/tsconfig.tsbuildinfo

src
tsconfig.json
vitest.config.ts