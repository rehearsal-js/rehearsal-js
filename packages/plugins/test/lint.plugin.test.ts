import { resolve } from 'path';
import { describe, expect, test } from 'vitest';
import { LintPlugin } from '../src/index.js';
import { initProject, mockPluginRunnerContext } from './utils.js';

describe('Test LintPlugin', () => {
  test('run', async () => {
    const project = await initProject('prettier-test', {
      'index.js': './test/fixtures/lint.ts.fixture',
    });

    const context = mockPluginRunnerContext(project);

    const plugin = new LintPlugin();

    for (const file in project.files) {
      const fileName = resolve(project.baseDir, file);

      const result = await plugin.run(fileName, context, {
        reportErrors: true,
        eslintOptions: {
          useEslintrc: false,
          fix: true,
          baseConfig: {
            extends: ['eslint:recommended', 'plugin:prettier/recommended'],
          },
        },
      });

      const resultText = context.service.getFileText(fileName).trim();

      expect(result).toHaveLength(1);
      expect(resultText).toMatchSnapshot();

      expect(context.reporter.currentRun.items).toHaveLength(1);
      expect(context.reporter.currentRun.items).toMatchSnapshot();
    }

    project.dispose();
  });
});
