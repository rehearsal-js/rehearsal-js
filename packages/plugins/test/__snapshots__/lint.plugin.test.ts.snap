// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Test LintPlugin > run 1`] = `
"function doSomething(arg) {
  return arg;
}"
`;

exports[`Test LintPlugin > run 2`] = `
[
  {
    "analysisTarget": "index.js",
    "category": "Error",
    "helpUrl": "",
    "hint": "'doSomething' is defined but never used.",
    "hintAdded": false,
    "message": "'doSomething' is defined but never used.",
    "nodeKind": "Identifier",
    "nodeLocation": {
      "endColumn": 21,
      "endLine": 1,
      "startColumn": 10,
      "startLine": 1,
    },
    "nodeText": "",
    "ruleId": "no-unused-vars",
    "type": 1,
  },
]
`;
